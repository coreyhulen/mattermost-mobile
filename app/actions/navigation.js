// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Navigation} from 'react-native-navigation';

import {getTheme} from 'mattermost-redux/selectors/entities/preferences';

export function resetToChannel() {
    return (dispatch, getState) => {
        const theme = getTheme(getState());

        Navigation.setRoot({
            root: {
                stack: {
                    children: [{
                        component: {
                            name: 'Channel',
                            options: {
                                layout: {
                                    backgroundColor: theme.centerChannelBg,
                                },
                            },
                        },
                    }],
                    options: {
                        statusBar: {
                            visible: true,
                        },
                        topBar: {
                            visible: false,
                        },
                    },
                },
            },
        });
    };
}

export function resetToSelectServer(allowOtherServers) {
    return () => {
        Navigation.setRoot({
            root: {
                stack: {
                    children: [{
                        component: {
                            name: 'SelectServer',
                            passProps: {
                                allowOtherServers,
                            },
                        },
                    }],
                    options: {
                        layout: {
                            backgroundColor: 'transparent',
                        },
                        statusBar: {
                            visible: true,
                        },
                        topBar: {
                            visible: false,
                        },
                    },
                },
            },
        });
    };
}
