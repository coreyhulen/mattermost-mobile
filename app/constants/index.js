// Copyright (c) 2016-present Mattermost, Inc. All Rights Reserved.
// See License.txt for license information.

import DeviceTypes from './device';
import NavigationTypes from './navigation';
import ViewTypes from './view';

export {
    DeviceTypes,
    NavigationTypes,
    ViewTypes
};
