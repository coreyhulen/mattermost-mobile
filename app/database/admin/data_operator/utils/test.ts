// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {mockedPosts} from './mock';
import {createPostsChain, sanitizePosts} from './utils';

describe('DataOperator: Utils tests', () => {
    it('sanitizePosts: should filter between ordered and unordered posts', () => {
        const {orderedPosts, unOrderedPosts} = sanitizePosts({posts: Object.values(mockedPosts.posts), orders: mockedPosts.order});
        expect(orderedPosts.length).toBe(4);
        expect(unOrderedPosts.length).toBe(2);
    });

    it('createPostsChain: should link posts amongst each other based on order array', () => {
        const previousPostId = 'prev_xxyuoxmehne';
        const chainedOfPosts = createPostsChain({orders: mockedPosts.order, rawPosts: Object.values(mockedPosts.posts), previousPostId});

        // eslint-disable-next-line max-nested-callbacks
        const post1 = chainedOfPosts.find((post) => {
            return post.id === '8swgtrrdiff89jnsiwiip3y1eoe';
        });
        expect(post1).toBeTruthy();
        expect(post1!.prev_post_id).toBe(previousPostId);

        // eslint-disable-next-line max-nested-callbacks
        const post2 = chainedOfPosts.find((post) => {
            return post.id === '8fcnk3p1jt8mmkaprgajoxz115a';
        });
        expect(post2).toBeTruthy();
        expect(post2!.prev_post_id).toBe('8swgtrrdiff89jnsiwiip3y1eoe');

        // eslint-disable-next-line max-nested-callbacks
        const post3 = chainedOfPosts.find((post) => {
            return post.id === '3y3w3a6gkbg73bnj3xund9o5ic';
        });
        expect(post3).toBeTruthy();
        expect(post3!.prev_post_id).toBe('8fcnk3p1jt8mmkaprgajoxz115a');

        // eslint-disable-next-line max-nested-callbacks
        const post4 = chainedOfPosts.find((post) => {
            return post.id === '4btbnmticjgw7ewd3qopmpiwqw';
        });
        expect(post4).toBeTruthy();
        expect(post4!.prev_post_id).toBe('3y3w3a6gkbg73bnj3xund9o5ic');
    });
});
