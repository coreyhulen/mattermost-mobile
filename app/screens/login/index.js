// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {bindActionCreators} from 'redux';
import {connect} from 'react-redux';

import {login} from 'mattermost-redux/actions/users';
import {getTheme} from 'mattermost-redux/selectors/entities/preferences';
import {getConfig, getLicense} from 'mattermost-redux/selectors/entities/general';
import {getColorScheme, isLandscape} from 'app/selectors/device';
import LoginActions from 'app/actions/views/login';
import {getColorStyles} from 'app/utils/appearance';

import Login from './login.js';

function mapStateToProps(state) {
    const {login: loginRequest} = state.requests.users;
    const config = getConfig(state);
    const colorScheme = getColorScheme(state);
    const colorStyles = getColorStyles(colorScheme);
    const license = getLicense(state);
    return {
        ...state.views.login,
        loginRequest,
        colorScheme,
        colorStyles,
        config,
        license,
        theme: getTheme(state),
        isLandscape: isLandscape(state),
    };
}

function mapDispatchToProps(dispatch) {
    return {
        actions: bindActionCreators({
            ...LoginActions,
            login,
        }, dispatch),
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(Login);
