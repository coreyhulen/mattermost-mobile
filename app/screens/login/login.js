// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {intlShape} from 'react-intl';
import {
    ActivityIndicator,
    Dimensions,
    Image,
    InteractionManager,
    Keyboard,
    StyleSheet,
    Text,
    TextInput,
    TouchableWithoutFeedback,
    View,
} from 'react-native';
import {Appearance} from 'react-native-appearance';
import Button from 'react-native-button';
import {KeyboardAwareScrollView} from 'react-native-keyboard-aware-scroll-view';
import {Navigation} from 'react-native-navigation';

import {paddingHorizontal as padding} from 'app/components/safe_area_view/iphone_x_spacing';
import ErrorText from 'app/components/error_text';
import FormattedText from 'app/components/formatted_text';
import StatusBar from 'app/components/status_bar';
import {resetToChannel, goToScreen} from 'app/actions/navigation';
import EphemeralStore from 'app/store/ephemeral_store';
import {
    getButtonStyle,
    getButtonTextStyle,
    getColorStyles,
    getInputStyle,
    getLogo,
    getStyledNavigationOptions,
} from 'app/utils/appearance';
import {preventDoubleTap} from 'app/utils/tap';
import tracker from 'app/utils/time_tracker';
import {t} from 'app/utils/i18n';
import {setMfaPreflightDone, getMfaPreflightDone} from 'app/utils/security';
import {GlobalStyles} from 'app/styles';

import telemetry from 'app/telemetry';

export const mfaExpectedErrors = ['mfa.validate_token.authenticate.app_error', 'ent.mfa.validate_token.authenticate.app_error'];

export default class Login extends PureComponent {
    static propTypes = {
        actions: PropTypes.shape({
            handleLoginIdChanged: PropTypes.func.isRequired,
            handlePasswordChanged: PropTypes.func.isRequired,
            handleSuccessfulLogin: PropTypes.func.isRequired,
            scheduleExpiredNotification: PropTypes.func.isRequired,
            login: PropTypes.func.isRequired,
        }).isRequired,
        config: PropTypes.object.isRequired,
        license: PropTypes.object.isRequired,
        loginId: PropTypes.string.isRequired,
        password: PropTypes.string.isRequired,
        isLandscape: PropTypes.bool.isRequired,
    };

    static contextTypes = {
        intl: intlShape.isRequired,
    };

    constructor(props) {
        super(props);

        this.state = {
            colorStyles: getColorStyles(Appearance.getColorScheme()),
            error: null,
            logo: getLogo(Appearance.getColorScheme()),
            isLoading: false,
        };
    }

    componentDidMount() {
        Dimensions.addEventListener('change', this.orientationDidChange);
        this.appearanceEventListener = Appearance.addChangeListener(({colorScheme}) => {
            const colorStyles = getColorStyles(colorScheme);
            this.setState({
                colorStyles,
                logo: getLogo(colorScheme),
            });

            Navigation.mergeOptions(EphemeralStore.getNavigationTopComponentId(), getStyledNavigationOptions(colorStyles));
        });

        setMfaPreflightDone(false);
    }

    componentWillUnmount() {
        Dimensions.removeEventListener('change', this.orientationDidChange);
        this.appearanceEventListener.remove();
    }

    goToChannel = () => {
        telemetry.remove(['start:overall']);

        tracker.initialLoad = Date.now();

        this.scheduleSessionExpiredNotification();

        resetToChannel();
    };

    goToMfa = () => {
        const {colorStyles} = this.state;
        const {intl} = this.context;
        const screen = 'MFA';
        const title = intl.formatMessage({id: 'mobile.routes.mfa', defaultMessage: 'Multi-factor Authentication'});

        goToScreen(screen, title, {onMfaComplete: this.checkLoginResponse}, getStyledNavigationOptions(colorStyles));
    };

    blur = () => {
        this.loginId.blur();
        this.passwd.blur();
        Keyboard.dismiss();
    };

    preSignIn = preventDoubleTap(() => {
        this.setState({error: null, isLoading: true});
        Keyboard.dismiss();
        InteractionManager.runAfterInteractions(async () => {
            if (!this.props.loginId) {
                t('login.noEmail');
                t('login.noEmailLdapUsername');
                t('login.noEmailUsername');
                t('login.noEmailUsernameLdapUsername');
                t('login.noLdapUsername');
                t('login.noUsername');
                t('login.noUsernameLdapUsername');

                // it's slightly weird to be constructing the message ID, but it's a bit nicer than triply nested if statements
                let msgId = 'login.no';
                if (this.props.config.EnableSignInWithEmail === 'true') {
                    msgId += 'Email';
                }
                if (this.props.config.EnableSignInWithUsername === 'true') {
                    msgId += 'Username';
                }
                if (this.props.license.IsLicensed === 'true' && this.props.config.EnableLdap === 'true') {
                    msgId += 'LdapUsername';
                }

                this.setState({
                    isLoading: false,
                    error: {
                        intl: {
                            id: msgId,
                            defaultMessage: '',
                            values: {
                                ldapUsername: this.props.config.LdapLoginFieldName ||
                                this.context.intl.formatMessage({
                                    id: 'login.ldapUsernameLower',
                                    defaultMessage: 'AD/LDAP username',
                                }),
                            },
                        },
                    },
                });
                return;
            }

            if (!this.props.password) {
                this.setState({
                    isLoading: false,
                    error: {
                        intl: {
                            id: t('login.noPassword'),
                            defaultMessage: 'Please enter your password',
                        },
                    },
                });
                return;
            }

            this.signIn();
        });
    });

    scheduleSessionExpiredNotification = () => {
        const {intl} = this.context;
        const {actions} = this.props;

        actions.scheduleExpiredNotification(intl);
    };

    signIn = () => {
        const {actions, loginId, password} = this.props;
        const {isLoading} = this.state;
        if (isLoading) {
            actions.login(loginId.toLowerCase(), password).
                then(this.checkLoginResponse);
        }
    };

    checkLoginResponse = (data) => {
        if (mfaExpectedErrors.includes(data?.error?.server_error_id)) { // eslint-disable-line camelcase
            this.goToMfa();
            this.setState({isLoading: false});
            return false;
        }

        if (data?.error) {
            this.setState({
                error: this.getLoginErrorMessage(data.error),
                isLoading: false,
            });
            return false;
        }

        this.setState({isLoading: false});
        resetToChannel();
        return true;
    };

    createLoginPlaceholder() {
        const {formatMessage} = this.context.intl;
        const license = this.props.license;
        const config = this.props.config;

        const loginPlaceholders = [];
        if (config.EnableSignInWithEmail === 'true') {
            loginPlaceholders.push(formatMessage({id: 'login.email', defaultMessage: 'Email'}));
        }

        if (config.EnableSignInWithUsername === 'true') {
            loginPlaceholders.push(formatMessage({id: 'login.username', defaultMessage: 'Username'}));
        }

        if (license.IsLicensed === 'true' && license.LDAP === 'true' && config.EnableLdap === 'true') {
            if (config.LdapLoginFieldName) {
                loginPlaceholders.push(config.LdapLoginFieldName);
            } else {
                loginPlaceholders.push(formatMessage({id: 'login.ldapUsername', defaultMessage: 'AD/LDAP Username'}));
            }
        }

        if (loginPlaceholders.length >= 2) {
            return loginPlaceholders.slice(0, loginPlaceholders.length - 1).join(', ') +
                ` ${formatMessage({id: 'login.or', defaultMessage: 'or'})} ` +
                loginPlaceholders[loginPlaceholders.length - 1];
        } else if (loginPlaceholders.length === 1) {
            return loginPlaceholders[0];
        }

        return '';
    }

    getLoginErrorMessage = (error) => {
        const errorMessage = this.getServerErrorForLogin(error) || this.state.error;
        if (errorMessage && this.loginId && this.passwd) {
            this.setErrorStyle();
        }
        return errorMessage;
    };

    getServerErrorForLogin = (error) => {
        if (!error) {
            return null;
        }
        const errorId = error.server_error_id;
        if (!errorId) {
            return error.message;
        }
        if (mfaExpectedErrors.includes(errorId) && !getMfaPreflightDone()) {
            return null;
        }
        if (
            errorId === 'store.sql_user.get_for_login.app_error' ||
            errorId === 'ent.ldap.do_login.user_not_registered.app_error'
        ) {
            return {
                intl: {
                    id: t('login.userNotFound'),
                    defaultMessage: "We couldn't find an account matching your login credentials.",
                },
            };
        } else if (
            errorId === 'api.user.check_user_password.invalid.app_error' ||
            errorId === 'ent.ldap.do_login.invalid_password.app_error'
        ) {
            return {
                intl: {
                    id: t('login.invalidPassword'),
                    defaultMessage: 'Your password is incorrect.',
                },
            };
        }
        return error.message;
    };

    loginRef = (ref) => {
        this.loginId = ref;
    };

    passwordRef = (ref) => {
        this.passwd = ref;
    };

    passwordFocus = () => {
        this.passwd.focus();
    };

    orientationDidChange = () => {
        this.scroll.scrollToPosition(0, 0, true);
    };

    scrollRef = (ref) => {
        this.scroll = ref;
    };

    forgotPassword = () => {
        const {colorStyles} = this.state;
        const {intl} = this.context;
        const screen = 'ForgotPassword';
        const title = intl.formatMessage({id: 'password_form.title', defaultMessage: 'Password Reset'});

        goToScreen(screen, title, {}, getStyledNavigationOptions(colorStyles));
    }

    isLoginButtonDisabled = () => {
        return !(this.props.loginId && this.props.password);
    };

    setErrorStyle() {
        this.setLoginStyle(GlobalStyles.inputBoxError);
        this.setPasswordStyle(GlobalStyles.inputBoxError);
    }

    setLoginStyle(style) {
        this.loginId.setNativeProps({style});
    }

    setPasswordStyle(style) {
        this.passwd.setNativeProps({style});
    }

    render() {
        const {colorStyles, logo, isLoading} = this.state;

        let proceed;
        if (isLoading) {
            proceed = (
                <ActivityIndicator
                    color={colorStyles.buttonTextDisabled.color}
                    animating={true}
                    size='small'
                />
            );
        } else {
            const additionalStyle = {};
            if (this.props.config.EmailLoginButtonColor) {
                additionalStyle.backgroundColor = this.props.config.EmailLoginButtonColor;
            }
            if (this.props.config.EmailLoginButtonBorderColor) {
                additionalStyle.borderColor = this.props.config.EmailLoginButtonBorderColor;
            }

            const additionalTextStyle = {};
            if (this.props.config.EmailLoginButtonTextColor) {
                additionalTextStyle.color = this.props.config.EmailLoginButtonTextColor;
            }

            proceed = (
                <Button
                    disabled={this.isLoginButtonDisabled()}
                    onPress={this.preSignIn}
                    containerStyle={[getButtonStyle(this.isLoginButtonDisabled(), colorStyles), additionalStyle]}
                >
                    <FormattedText
                        id='login.signIn'
                        defaultMessage='Sign in'
                        style={[getButtonTextStyle(this.isLoginButtonDisabled(), colorStyles), additionalTextStyle]}
                    />
                </Button>
            );
        }

        let forgotPassword;
        if (this.props.config.EnableSignInWithEmail === 'true' || this.props.config.EnableSignInWithUsername === 'true') {
            forgotPassword = (
                <Button
                    onPress={this.forgotPassword}
                    containerStyle={[style.forgotPasswordBtn]}
                >
                    <FormattedText
                        id='login.forgot'
                        defaultMessage='I forgot my password'
                        style={colorStyles.link}
                    />
                </Button>
            );
        }

        return (
            <View style={[GlobalStyles.container, colorStyles.container]}>
                <StatusBar/>
                <TouchableWithoutFeedback
                    onPress={this.blur}
                    accessible={false}
                >
                    <KeyboardAwareScrollView
                        ref={this.scrollRef}
                        style={[GlobalStyles.container, colorStyles.container]}
                        contentContainerStyle={[GlobalStyles.innerContainer, padding(this.props.isLandscape)]}
                        keyboardShouldPersistTaps='handled'
                        enableOnAndroid={true}
                    >
                        <Image
                            source={logo}
                        />
                        <View>
                            <Text style={[GlobalStyles.header, colorStyles.header]}>
                                {this.props.config.SiteName}
                            </Text>
                            <FormattedText
                                style={[GlobalStyles.subheader, colorStyles.header]}
                                id='web.root.signup_info'
                                defaultMessage='All team communication in one place, searchable and accessible anywhere'
                            />
                        </View>
                        <TextInput
                            ref={this.loginRef}
                            value={this.props.loginId}
                            onBlur={this.setLoginStyle.bind(this, colorStyles.inputBox)}
                            onChangeText={this.props.actions.handleLoginIdChanged}
                            onFocus={this.setLoginStyle.bind(this, colorStyles.inputBoxFocused)}
                            style={getInputStyle(isLoading, colorStyles)}
                            placeholder={this.createLoginPlaceholder()}
                            placeholderTextColor={colorStyles.inputBoxDisabled.color}
                            autoCorrect={false}
                            autoCapitalize='none'
                            keyboardType='email-address'
                            returnKeyType='next'
                            underlineColorAndroid='transparent'
                            onSubmitEditing={this.passwordFocus}
                            blurOnSubmit={false}
                            disableFullscreenUI={true}
                            editable={!isLoading}
                        />
                        <TextInput
                            ref={this.passwordRef}
                            value={this.props.password}
                            onBlur={this.setPasswordStyle.bind(this, colorStyles.inputBox)}
                            onChangeText={this.props.actions.handlePasswordChanged}
                            onFocus={this.setPasswordStyle.bind(this, colorStyles.inputBoxFocused)}
                            style={[GlobalStyles.inputBox, colorStyles.inputBox]}
                            placeholder={this.context.intl.formatMessage({id: 'login.password', defaultMessage: 'Password'})}
                            placeholderTextColor={colorStyles.inputBoxDisabled.color}
                            secureTextEntry={true}
                            autoCorrect={false}
                            autoCapitalize='none'
                            underlineColorAndroid='transparent'
                            returnKeyType='go'
                            onSubmitEditing={this.preSignIn}
                            disableFullscreenUI={true}
                            editable={!isLoading}
                        />
                        <ErrorText error={this.state.error}/>
                        {proceed}
                        {forgotPassword}
                    </KeyboardAwareScrollView>
                </TouchableWithoutFeedback>
            </View>
        );
    }
}

const style = StyleSheet.create({
    forgotPasswordBtn: {
        borderColor: 'transparent',
        marginTop: 15,
    },
});
