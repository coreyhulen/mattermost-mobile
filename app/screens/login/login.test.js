// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {Appearance} from 'react-native-appearance';

import {RequestStatus} from 'mattermost-redux/constants';

import FormattedText from 'app/components/formatted_text';

import {shallowWithIntl} from 'test/intl-test-helper';

import * as NavigationActions from 'app/actions/navigation';

import {mfaExpectedErrors} from 'app/screens/login/login';
import {darkColors, lightColors} from 'app/styles/colors';
import Login from './login';
import {getColorStyles, getStyledNavigationOptions} from '../../utils/appearance';

describe('Login', () => {
    const baseProps = {
        config: {
            EnableSignInWithEmail: 'true',
            EnableSignInWithUsername: 'true',
        },
        license: {
            IsLicensed: 'false',
        },
        loginId: '',
        password: '',
        loginRequest: {},
        actions: {
            handleLoginIdChanged: jest.fn(),
            handlePasswordChanged: jest.fn(),
            handleSuccessfulLogin: jest.fn(),
            scheduleExpiredNotification: jest.fn(),
            login: jest.fn(),
        },
        isLandscape: false,
    };

    test('should show "I forgot my password" with only email login enabled', () => {
        const props = {
            ...baseProps,
            config: {
                ...baseProps.config,
                EnableSignInWithUsername: 'false',
            },
        };

        const wrapper = shallowWithIntl(<Login {...props}/>);

        expect(wrapper.find(FormattedText).find({id: 'login.forgot'}).exists()).toBe(true);
    });

    test('should show "I forgot my password" with only username login enabled', () => {
        const props = {
            ...baseProps,
            config: {
                ...baseProps.config,
                EnableSignInWithEmail: 'false',
            },
        };

        const wrapper = shallowWithIntl(<Login {...props}/>);

        expect(wrapper.find(FormattedText).find({id: 'login.forgot'}).exists()).toBe(true);
    });

    test('should not show "I forgot my password" without email or username login enabled', () => {
        const props = {
            ...baseProps,
            config: {
                ...baseProps.config,
                EnableSignInWithEmail: 'false',
                EnableSignInWithUsername: 'false',
            },
        };

        const wrapper = shallowWithIntl(<Login {...props}/>);

        expect(wrapper.find(FormattedText).find({id: 'login.forgot'}).exists()).toBe(false);
    });

    test('should send the user to the login screen after login', (done) => {
        const resetToChannel = jest.spyOn(NavigationActions, 'resetToChannel').mockImplementation(() => done());

        let props = {
            ...baseProps,
            loginRequest: {
                status: RequestStatus.NOT_STARTED,
            },
        };

        props.actions.handleSuccessfulLogin.mockImplementation(() => Promise.resolve());

        const wrapper = shallowWithIntl(<Login {...props}/>);

        expect(resetToChannel).not.toHaveBeenCalled();

        props = {
            ...props,
            loginRequest: {
                status: RequestStatus.STARTED,
            },
        };
        wrapper.setProps(props);

        expect(resetToChannel).not.toHaveBeenCalled();

        props = {
            ...props,
            loginRequest: {
                status: RequestStatus.SUCCESS,
            },
        };
        wrapper.setProps(props);

        // This test times out if resetToChannel hasn't been called
    });

    test('should go to MFA screen when login response returns MFA error', () => {
        const goToScreen = jest.spyOn(NavigationActions, 'goToScreen');

        const mfaError = {
            error: {
                server_error_id: mfaExpectedErrors[0],
            },
        };

        const wrapper = shallowWithIntl(<Login {...baseProps}/>);
        wrapper.instance().checkLoginResponse(mfaError);

        expect(goToScreen).
            toHaveBeenCalledWith(
                'MFA',
                'Multi-factor Authentication',
                {},
                getStyledNavigationOptions(getColorStyles('dark'))
            );
    });

    test('should go to ForgotPassword screen when forgotPassword is called', () => {
        const goToScreen = jest.spyOn(NavigationActions, 'goToScreen');

        const wrapper = shallowWithIntl(<Login {...baseProps}/>);
        wrapper.instance().forgotPassword();

        expect(goToScreen).
            toHaveBeenCalledWith(
                'ForgotPassword',
                'Password Reset',
                {},
                getStyledNavigationOptions(getColorStyles('dark'))
            );
    });

    test('should show light background when user has light color scheme set', () => {
        Appearance.getColorScheme.mockImplementation(() => 'light');
        Appearance.addChangeListener.mockImplementation(() => 'light');

        const wrapper = shallowWithIntl(<Login {...baseProps}/>);

        expect(wrapper.getElement()).toMatchSnapshot();
        expect(wrapper).toHaveStyle('backgroundColor', lightColors.containerBg);
    });

    test('should show dark background when user has dark color scheme set', () => {
        Appearance.getColorScheme.mockImplementation(() => 'dark');
        Appearance.addChangeListener.mockImplementation(() => 'dark');

        const wrapper = shallowWithIntl(<Login {...baseProps}/>);

        expect(wrapper.getElement()).toMatchSnapshot();
        expect(wrapper).toHaveStyle('backgroundColor', darkColors.containerBg);
    });
});
