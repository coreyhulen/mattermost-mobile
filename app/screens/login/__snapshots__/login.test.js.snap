// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should show dark background when user has dark color scheme set 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "backgroundColor": "#171717",
      },
    ]
  }
>
  <Connect(StatusBar) />
  <TouchableWithoutFeedback
    onPress={[Function]}
  >
    <KeyboardAwareScrollViewMock
      contentContainerStyle={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "column",
            "justifyContent": "center",
            "paddingHorizontal": 15,
            "paddingVertical": 50,
          },
          null,
        ]
      }
      enableAutomaticScroll={true}
      enableOnAndroid={true}
      enableResetScrollToCoords={true}
      extraHeight={75}
      extraScrollHeight={0}
      keyboardOpeningTime={250}
      keyboardShouldPersistTaps="handled"
      style={
        Array [
          Object {
            "flex": 1,
          },
          Object {
            "backgroundColor": "#171717",
          },
        ]
      }
      viewIsInsideTabBar={false}
    >
      <Image
        source={
          Object {
            "testUri": "../../../dist/assets/images/logo_dark.png",
          }
        }
      />
      <View>
        <Text
          style={
            Array [
              Object {
                "fontSize": 32,
                "fontWeight": "600",
                "marginBottom": 15,
                "marginTop": 15,
                "textAlign": "center",
              },
              Object {
                "color": "#dbdbdb",
              },
            ]
          }
        />
        <FormattedText
          defaultMessage="All team communication in one place, searchable and accessible anywhere"
          id="web.root.signup_info"
          style={
            Array [
              Object {
                "color": "#777",
                "fontSize": 16,
                "fontWeight": "300",
                "lineHeight": 22,
                "marginBottom": 15,
                "textAlign": "center",
              },
              Object {
                "color": "#dbdbdb",
              },
            ]
          }
        />
      </View>
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        autoCorrect={false}
        blurOnSubmit={false}
        disableFullscreenUI={true}
        editable={true}
        keyboardType="email-address"
        onBlur={[Function]}
        onChangeText={[MockFunction]}
        onFocus={[Function]}
        onSubmitEditing={[Function]}
        placeholder="Email or Username"
        placeholderTextColor="#9b9b9b"
        rejectResponderTermination={true}
        returnKeyType="next"
        style={
          Array [
            Object {
              "alignSelf": "stretch",
              "borderRadius": 3,
              "borderWidth": 2,
              "color": "#3d3c40",
              "fontSize": 16,
              "height": 45,
              "marginBottom": 5,
              "marginTop": 5,
              "paddingLeft": 10,
            },
            Object {
              "borderColor": "#9b9b9b",
              "color": "#dbdbdb",
            },
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        autoCorrect={false}
        disableFullscreenUI={true}
        editable={true}
        onBlur={[Function]}
        onChangeText={[MockFunction]}
        onFocus={[Function]}
        onSubmitEditing={[Function]}
        placeholder="Password"
        placeholderTextColor="#9b9b9b"
        rejectResponderTermination={true}
        returnKeyType="go"
        secureTextEntry={true}
        style={
          Array [
            Object {
              "alignSelf": "stretch",
              "borderRadius": 3,
              "borderWidth": 2,
              "color": "#3d3c40",
              "fontSize": 16,
              "height": 45,
              "marginBottom": 5,
              "marginTop": 5,
              "paddingLeft": 10,
            },
            Object {
              "borderColor": "#9b9b9b",
              "color": "#dbdbdb",
            },
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
      <Connect(ErrorText)
        error={null}
      />
      <Button
        containerStyle={
          Array [
            Object {
              "alignItems": "center",
              "alignSelf": "stretch",
              "borderRadius": 3,
              "marginTop": 10,
              "padding": 15,
            },
            Object {
              "backgroundColor": "#399fff",
            },
            Object {
              "backgroundColor": "#2e2e2e",
            },
          ]
        }
        disabled={true}
        onPress={[Function]}
      >
        <FormattedText
          defaultMessage="Sign in"
          id="login.signIn"
          style={
            Array [
              Object {
                "fontSize": 17,
                "fontWeight": "500",
                "textAlign": "center",
              },
              Object {
                "color": "#171717",
              },
              Object {
                "color": "#9b9b9b",
              },
            ]
          }
        />
      </Button>
      <Button
        containerStyle={
          Array [
            Object {
              "borderColor": "transparent",
              "marginTop": 15,
            },
          ]
        }
        onPress={[Function]}
      >
        <FormattedText
          defaultMessage="I forgot my password"
          id="login.forgot"
          style={
            Object {
              "color": "#399fff",
            }
          }
        />
      </Button>
    </KeyboardAwareScrollViewMock>
  </TouchableWithoutFeedback>
</View>
`;

exports[`Login should show light background when user has light color scheme set 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "backgroundColor": "#ffffff",
      },
    ]
  }
>
  <Connect(StatusBar) />
  <TouchableWithoutFeedback
    onPress={[Function]}
  >
    <KeyboardAwareScrollViewMock
      contentContainerStyle={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "column",
            "justifyContent": "center",
            "paddingHorizontal": 15,
            "paddingVertical": 50,
          },
          null,
        ]
      }
      enableAutomaticScroll={true}
      enableOnAndroid={true}
      enableResetScrollToCoords={true}
      extraHeight={75}
      extraScrollHeight={0}
      keyboardOpeningTime={250}
      keyboardShouldPersistTaps="handled"
      style={
        Array [
          Object {
            "flex": 1,
          },
          Object {
            "backgroundColor": "#ffffff",
          },
        ]
      }
      viewIsInsideTabBar={false}
    >
      <Image
        source={
          Object {
            "testUri": "../../../dist/assets/images/logo_light.png",
          }
        }
      />
      <View>
        <Text
          style={
            Array [
              Object {
                "fontSize": 32,
                "fontWeight": "600",
                "marginBottom": 15,
                "marginTop": 15,
                "textAlign": "center",
              },
              Object {
                "color": "#171717",
              },
            ]
          }
        />
        <FormattedText
          defaultMessage="All team communication in one place, searchable and accessible anywhere"
          id="web.root.signup_info"
          style={
            Array [
              Object {
                "color": "#777",
                "fontSize": 16,
                "fontWeight": "300",
                "lineHeight": 22,
                "marginBottom": 15,
                "textAlign": "center",
              },
              Object {
                "color": "#171717",
              },
            ]
          }
        />
      </View>
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        autoCorrect={false}
        blurOnSubmit={false}
        disableFullscreenUI={true}
        editable={true}
        keyboardType="email-address"
        onBlur={[Function]}
        onChangeText={[MockFunction]}
        onFocus={[Function]}
        onSubmitEditing={[Function]}
        placeholder="Email or Username"
        placeholderTextColor="#5c5c5c"
        rejectResponderTermination={true}
        returnKeyType="next"
        style={
          Array [
            Object {
              "alignSelf": "stretch",
              "borderRadius": 3,
              "borderWidth": 2,
              "color": "#3d3c40",
              "fontSize": 16,
              "height": 45,
              "marginBottom": 5,
              "marginTop": 5,
              "paddingLeft": 10,
            },
            Object {
              "borderColor": "#dbdbdb",
              "color": "#171717",
            },
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        autoCorrect={false}
        disableFullscreenUI={true}
        editable={true}
        onBlur={[Function]}
        onChangeText={[MockFunction]}
        onFocus={[Function]}
        onSubmitEditing={[Function]}
        placeholder="Password"
        placeholderTextColor="#5c5c5c"
        rejectResponderTermination={true}
        returnKeyType="go"
        secureTextEntry={true}
        style={
          Array [
            Object {
              "alignSelf": "stretch",
              "borderRadius": 3,
              "borderWidth": 2,
              "color": "#3d3c40",
              "fontSize": 16,
              "height": 45,
              "marginBottom": 5,
              "marginTop": 5,
              "paddingLeft": 10,
            },
            Object {
              "borderColor": "#dbdbdb",
              "color": "#171717",
            },
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
      <Connect(ErrorText)
        error={null}
      />
      <Button
        containerStyle={
          Array [
            Object {
              "alignItems": "center",
              "alignSelf": "stretch",
              "borderRadius": 3,
              "marginTop": 10,
              "padding": 15,
            },
            Object {
              "backgroundColor": "#1247b1",
            },
            Object {
              "backgroundColor": "#dbdbdb",
            },
          ]
        }
        disabled={true}
        onPress={[Function]}
      >
        <FormattedText
          defaultMessage="Sign in"
          id="login.signIn"
          style={
            Array [
              Object {
                "fontSize": 17,
                "fontWeight": "500",
                "textAlign": "center",
              },
              Object {
                "color": "#ffffff",
              },
              Object {
                "color": "#5c5c5c",
              },
            ]
          }
        />
      </Button>
      <Button
        containerStyle={
          Array [
            Object {
              "borderColor": "transparent",
              "marginTop": 15,
            },
          ]
        }
        onPress={[Function]}
      >
        <FormattedText
          defaultMessage="I forgot my password"
          id="login.forgot"
          style={
            Object {
              "color": "#1247b1",
            }
          }
        />
      </Button>
    </KeyboardAwareScrollViewMock>
  </TouchableWithoutFeedback>
</View>
`;
